core/src/server/mod.rs

```diff
@@ -42,3 +42,74 @@ pub use helpers::{BatchResponseBuilder, BoundedWriter, MethodResponse, MethodSin
 pub use host_filtering::*;
 pub use rpc_module::*;
 pub use subscription::*;
+
+use jsonrpsee_types::{ErrorObjectOwned, PartialResponse};
+
+/// Something that can be converted into a JSON-RPC error object.
+pub trait IntoErrorObject {
+	/// Something that can be converted into a JSON-RPC error object.
+	fn into_error_object(self) -> ErrorObjectOwned;
+}
+
+impl IntoErrorObject for crate::Error {
+	fn into_error_object(self) -> ErrorObjectOwned {
+		self.into()
+	}
+}
+
+/// Something that can be converted into a JSON-RPC method call response.
+///
+/// If the value couldn't be serialized/encoded, jsonrpsee will sent out an error
+/// to the client `response could not be serialized`.
+pub trait IntoResponse {
+	/// Output.
+	type Output: serde::Serialize;
+
+	/// Something that can be converted into a JSON-RPC method call response.
+	fn into_response(self) -> PartialResponse<Self::Output>;
```

Comment:
I guess I'm wondering ; can we justify the differences between this and 

```rust
pub trait IntoSubscriptionCloseResponse {
	fn into_response(self) -> SubscriptionCloseResponse;
}
```

Right now `IntoResponse` is really quite different (more generic, for instance with its associated param, and so on). Should these traits be aiming to be as consistent with eachother as makes sense, or are there good reasons for the differences that we can explain?

Reply:
I would say that they are fundamentally different mainly because `IntoResponse` is something that converts a response to a JSON-RPC response or JSON-RPC error object and `IntoSubscriptionClose` is really just something that is converted into a subscription notification.

In practice, users would not need to implement `IntoResponse` except some rare edge-cases for custom types that are not `Result<T, impl IntoErrorObject`. For example you could return `enum MyResponse { Success(String), Error1, Error }` where one would need to implement this trait

Thus, is in practice the only trait that folks needs to implement is `IntoErrorObject` which a specific error code and so on.

Reply:
> I would say that they are fundamentally different mainly because IntoResponse is something that converts a response to a JSON-RPC response or JSON-RPC error object (a response with the result or error entry). IntoSubscriptionClose is really just something that is converted into a subscription notification.

I guess what I meant was that they both basically take some type and return some part of a JSON response, but this one comes with an associated type and is more generic, and this isn't. It's not a huge deal for me; just wondering whether there is anything that should be made more consistent between these traits, but happy if not :)

Reply:
we want to enforce these formats and be flexible enough for folks to use:

```json
{"jsonrpc": "2.0", "result": -19, "id": 2}
```

or 

```json
{"jsonrpc": "2.0", "error": {"code": -32601, "message": "Method not found"}, "id": "1"}
```

so we could serialize the response in the `IntoResponse impl` to remove the generic stuff which probably is better

Reply:
I guess efficiency is the main thing; if we can be less generic ie have something like:

```rust
pub trait IntoResponse {
	fn into_response(self) -> Response; // returns some hardcoded Response type
}
```

Where `into_response` is serializing into some non-generic type, and _if_ that doesn't lose out in efficiency (because we have to serializie stuff anyway or whatever) then I geuss it'd be easier to udnerstand and more consistent with the subscription version.

But if having the generic `PartialResponse<Self::Output>` type thing makes stuff more efficient then I get it :)